cranelift/codegen/src/ir/pcc.rs

```diff
@@ -479,6 +544,293 @@ impl fmt::Display for Fact {
+            max: smallvec![max],
+        }
+    }
+
+    /// Create a range that is exactly one expression, with another expression as an upper bound.
+    pub fn exact_with_max(expr: Expr, max: Expr) -> Self {
+        ValueRange {
+            equal: smallvec![expr],
+            min: smallvec![],
+            max: smallvec![max],
+        }
+    }
+
+    /// Is this ValueRange an exact integer constant?
+    pub fn as_const(&self) -> Option<i128> {
+        self.equal.iter().find_map(|&e| e.as_const())
+    }
+
+    /// Is this ValueRange definitely less than or equal to the given expression?
+    pub fn le_expr(&self, expr: &Expr) -> bool {
+        // The range is <= the expr if *any* of its upper bounds are
+        // <= the expr, because each upper bound constrains the whole
+        // range (i.e., the range is the intersection of all
+        // combinations of bounds). Likewise, if any expression that
+        // exactly determines the value less than `expr`, then we can
+        // definitely say the range is less than `expr`.
+        let result = self
+            .equal
+            .iter()
+            .chain(self.max.iter())
+            .any(|e| Expr::le(e, expr));
+        trace!("ValueRange::le_expr: {self:?} {expr:?} -> {result}");
+        result
+    }
+
+    /// Is the expression definitely within the ValueRange?
+    pub fn contains_expr(&self, expr: &Expr) -> bool {
+        let result = ((!self.min.is_empty() || !self.max.is_empty())
+            && self
+                .min
+                .iter()
+                .all(|lower_bound| Expr::le(lower_bound, expr))
+            && self
+                .max
+                .iter()
+                .all(|upper_bound| Expr::le(expr, upper_bound)))
+            || self.equal.iter().any(|equiv| equiv == expr);
+        trace!("ValueRange::contains_expr: {self:?} {expr:?} -> {result}");
+        result
+    }
+
+    /// Simplify a ValueRange by removing redundant bounds. Any lower
+    /// bound greater than another lower bound, or any upper bound
+    /// less than another upper bound, can be removed.
+    pub fn simplify(&mut self) {
+        trace!("simplify: {self:?}");
+
+        // Note an important invariant: syntactic equality of Exprs
+        // implies symbolic equality. This is required to ensure we
+        // don't remove both `x` and `y` if `x <= y` and `y <= x`,
+        // given the logic below.
+        self.equal.sort();
+        self.equal.dedup();
+
+        // A lower bound `e` is not redundant if for all other
+        // lower bounds `other`, we cannot show that `e >=
+        // other`.
+        self.min.sort();
+        self.min.dedup();
+        let min = self
+            .min
+            .iter()
+            .filter(|&e| {
+                self.min
+                    .iter()
+                    .all(|other| e == other || !Expr::le(other, e))
+            })
+            .cloned()
+            .collect::<SmallVec<[Expr; 1]>>();
```

Comment:
The `O(n^2)` loop is a little scary here but I guess in practice this is sized at most by the number of unique values we GVN together? Which seems like it should generally be relatively small? But also something that is ~unbounded and user-controllable :grimacing: 

But I guess they aren't controlling the actual facts used during a wasm load and the inputs to an `i32.load` instruction are basically any `i32` value. So I think this is maybe okay?

Seems like something we should think about and have a comment for calming readers down.

Reply:
So, this is a very interesting point: I was able to control the number of `equal` expressions by writing a bunch of addresses that simplify to the same value. (E.g., loads from `p` and `p + (4-4)`.) Lower and upper bounds on the other hand are not user-controllable as the dynamic ones arise from memories and any given access only touches one memory (and static ones are simplified so that only one lower and one upper remain). Great catch and thank you for thinking of this!

The need for "equal" constraints arises from the transitivity -- we need to represent the values `v1` and `gv1` when we see `icmp gte v1, gv1` -- but I suspect we might be able to get away with at most one. This does mean we lose information sometimes but...

(A possible sketch of a strategy: when we merge, always keep the "lesser" expression per the arbitrary lexicographical ordering; as long as we're consistent, we keep the comparison info for `v1` vs. `gv1` but not `v2` vs. `gv1`, across all related facts, and all related facts should merge if we have two loads of addresses `v1` and `v2` that ended up merging during opt.)

Reply:
> (A possible sketch of a strategy: when we merge, always keep the "lesser" expression per the arbitrary lexicographical ordering; as long as we're consistent, we keep the comparison info for `v1` vs. `gv1` but not `v2` vs. `gv1`, across all related facts, and all related facts should merge if we have two loads of addresses `v1` and `v2` that ended up merging during opt.)

This sounds very reasonable to me.